File: ruby/user_management_db_adapter/lib/user_management/in_memory_user_repository.rb
require 'user_management/user_repository'

module UserManagement
  class InMemoryUserRepository
    include UserRepository

    def initialize
      @store = {}
    end

    def save(user)
      @store[user.id] = user
    end

    def find_by_id(user_id)
      @store[user_id]
    end

    def find_all
      @store.values
    end

    def delete(user_id)
      @store.delete(user_id)
    end
  end
end

File: java/user-management-db-adapter/src/main/java/com/hexagonalchess/usermanagement/InMemoryUserRepository.java
package com.hexagonalchess.usermanagement;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class InMemoryUserRepository implements UserRepository {

    private final Map<UserId, User> store = new LinkedHashMap<>();

    @Override
    public void save(User user) {
        store.put(user.id(), user);
    }

    @Override
    public Optional<User> findById(UserId id) {
        return Optional.ofNullable(store.get(id));
    }

    @Override
    public List<User> findAll() {
        return new ArrayList<>(store.values());
    }

    @Override
    public void delete(UserId id) {
        store.remove(id);
    }
}
